[#64] 좌석 구분 및 가격 지정

tlsgkr7416 opened the pull request:
1. PerformanceController
    * 공연 정보 추가 시 좌석 구분 및 가격 정보도 추가
2. SeatPriceService
    * checkStartSeat 메서드로 좌석 시작 번호가 적절한지 체크
    * checkEndSeat 메서드로 좌석 마지막 번호가 공연 홀 좌석 수보다 큰지 체크(공연 홀의 총 좌석 수를 venueHallService를 통해 받아옴)
    * checkDuplicationSeatNumber 메서드로 유저가 SeatPriceRequest를 List 형식으로 여러 개 보내는데
      SeatPriceRequest에 시작 좌석과 마지막 좌석 중 다른 SeatPriceRequest의 좌석과 겹치는 게 있는지 체크
    * insertSeatsPrice 메서드로 좌석 구분 및 가격 정보 insert 로직
3. venueHallService
    * 공연 홀의 총 좌석 수를 db에서 받아 오는 로직 추가
4. RatingType
    * 좌석 등급 /1 VIP/ /2 S/ /3 A/ enum 처리
5. PerformanceRequest
    * 공연 정보를 추가 후 바로 DB에서 id를 받아와 좌석 구분 및 가격 지정 로직을 처리하기 위해 PerformanceRequest에 Id 추가
6. 좌석 구분 및 가격 정보 SQL
    * seatPrice 테이블

tlsgkr7416 committed:
Merge remote-tracking branch 'origin/feature/34/jh' into feature/34/jh

tlsgkr7416 committed:
[#64] 좌석 구분 및 가격 지정

1. PerformanceController
    * 공연 정보 추가 시 좌석 구분 및 가격 정보도 추가
2. SeatPriceService
    * checkStartSeat 메서드로 좌석 시작 번호가 적절한지 체크
    * checkEndSeat 메서드로 좌석 마지막 번호가 공연 홀 좌석 수보다 큰지 체크(공연 홀의 총 좌석 수를 venueHallService를 통해 받아옴)
    * checkDuplicationSeatNumber 메서드로 유저가 SeatPriceRequest를 List 형식으로 여러 개 보내는데
      SeatPriceRequest에 시작 좌석과 마지막 좌석 중 다른 SeatPriceRequest의 좌석과 겹치는 게 있는지 체크
    * insertSeatsPrice 메서드로 좌석 구분 및 가격 정보 insert 로직
3. venueHallService
    * 공연 홀의 총 좌석 수를 db에서 받아 오는 로직 추가
4. RatingType
    * 좌석 등급 /1 VIP/ /2 S/ /3 A/ enum 처리
5. PerformanceRequest
    * 공연 정보를 추가 후 바로 DB에서 id를 받아와 좌석 구분 및 가격 지정 로직을 처리하기 위해 PerformanceRequest에 Id 추가
6. 좌석 구분 및 가격 정보 SQL
    * seatPrice 테이블

hanwix2 commented:
관련된 이슈들만 PR에 포함되면 좋을 것 같아요~

hanwix2 commented on src/main/java/com/show/showticketingservice/model/performance/SeatPriceRequest.java:
```diff
@@ -0,0 +1,25 @@
+    private final int price;
+
+    @NotNull(message = "등급을 입력하세요.")
+    private final int rating;
```
`RatingType` Enum을 사용해야 하지 않을까요?

hanwix2 commented on src/main/java/com/show/showticketingservice/service/SeatPriceService.java:
```diff
@@ -0,0 +1,63 @@
+        checkDuplicationSeatNumber(seatPriceRequests);
+
+        seatPriceMapper.insertSeatsPrice(seatPriceRequests, performanceId);
+    }
```
1. 좌석 등급이 **최대 3개**인데 이것을 체크하는 로직도 필요할 것 같습니다.
2. 등급이 중복된 경우는 체크하지만 만약 **등급 지정이 안된 경우**도 체크해야하지 않을까요? 모든 자리에 가격 정보가 필요하니 누락되지 않아야 할 것 같습니다.
3. 좌석을 번호로 구분하면 등급을 구분하기 어려울 것 같습니다. **행으로 구분**하는 것은 어떨까요?
4. 보통 앞 좌석일수록 등급이 높은데, 이것을 관리자가 입력한 좌석 등급 수에 맞게 **알맞은 등급을 자동으로 지정**하면 좋지 않을까요? 1, 2, 3번 모두 연관된 기능일 것 같은데 오류의 가능성을 최소화 하는게 좋을 것 같습니다.

f-lab-dev commented:
정한님이 말씀하신대로 너무 많은 이슈에 대한 커밋로그가 있네요. 타겟 브랜치 조정이 필요해보입니다~

tlsgkr7416 committed:
[#64] 공연 좌석 가격 구분 기준을 행으로 변경

1. VenueHallMapper
    * 공연Id로 공연 홀의 총 행 좌석을 갖고 오기 위해 공연 홀과 공연 정보를 join 하여 총 행 좌석을 갖고 옴
2. SeatPriceService의 메서드
    * checkStartColNum : 좌석 시작 행이 마지막 좌석 행보다 클 경우 예외 처리
    * checkEndColNum : 마지막 좌석 행이 공연 홀 총 좌석 행 보다 클 경우 예외 처리
    * checkDuplicationColNum : 좌석 행을 추가 시 겹치는 행이 있을 경우 예외 처리
    * checkEmptyColNum : 비어있는 좌석 행이 있을 시 예외 처리
    * checkDuplicationSeatsRatingList : List<SeatPriceRequest>로 등급을 VIP, S, A 중 선택하여 보낼 수 있음(단 중복은 제외)
     만약 List<SeatPriceRequest> VIP, VIP 두개로 보낼 경우 중복되어 예외 처리. 따라서 등급을 중복하여 보내는 게 불가능해 최대 3개까지 보낼 수 있게 해줌.
3. PerformanceController
    * insertPerfPrice로 좌석 가격 handler 분리
3. 테스트 코드 작성
4. 기능
    * 좌석의 등급은 행을 기준으로 보낼 수 있음
    * 클라이언트는 좌석 등급을 VIP, S, A 중 선택해서 보내야 하고 등급을 하나만 선택해서 보내도 됨(단 중복은 제외)
    * 등급 순서는 VIP, S, A 로 앞좌석 부터 지정
    * 행을 1줄만 등급을 지정 하고 싶다면 startColNum 과 endColNum을 같은 행으로 보내면 됨

hanwix2 commented on etc/show_ticketing_service.sql:
```diff
@@ -74,9 +74,9 @@ CREATE TABLE seatPrice
-    `startSeat`      INT             NOT NULL,
-    `endSeat`        INT             NOT NULL,
+    `ratingType`     INT             NOT NULL,
+    `startColNum`    INT             NOT NULL,
+    `endColNum`      INT             NOT NULL,
```
행으로 구분한다고 하셨는데 **RowNum**이 아닐까요?

tlsgkr7416 replied:
영어 단어를 착각했네요.

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/LocalFileService.java:
```diff
@@ -0,0 +1,65 @@
+    private String fileUploadPath;
+
+    @Override
+    public String registerPosterImage(MultipartFile image) {
```
이 부분은 정한님의 작업과 겹치는 것 같은데 타겟브랜치가 잘못 설정된 것 같습니다~

tlsgkr7416 replied:
현재 타겟 브랜치가 develop밖에 없는데 develop에 있는 파일을 pull 해서 충돌을 일단 해결해야 할 거 같네요....

hanwix2 commented:
`SeatPrice.xml` 파일 명에 **mapper**를 붙여 다른 매핑 xml파일과 통일 시켜주시면 좋을 듯 합니다.

tlsgkr7416 committed:
[#64] 좌석 구분 및 가격 지정 수정

1. colNum => rowNum 수정
2. xml 파일명 수정

tlsgkr7416 committed:
Merge remote-tracking branch 'origin/develop' into feature/64/sh

tlsgkr7416 committed:
[#64] 좌석 구분 및 가격 지정 수정

1. 공연 좌석 가격이 이미 db에 존재 시 예외처리 기능 추가
2. PerformanceController
    * insertPerfPrice에 list로 요청을 하여 @Validated 추가

tlsgkr7416 committed:
[#64] 공연 좌석 가격 구분 수정

1. @valid 위치 수정
2. 메소드명 수정
3. 예외 처리 수정

f-lab-dev commented on src/main/java/com/show/showticketingservice/exception/performance/SameSeatRatingListAdditionException.java:
```diff
@@ -0,0 +1,7 @@
+package com.show.showticketingservice.exception.performance;
+
+public class SameSeatRatingListAdditionException extends RuntimeException {
```
여기도 아래 공백 하나 있으면 좋을 것 같네요~

tlsgkr7416 committed:
[#64] 좌석 구분 및 가격 피드백 적용

f-lab-dev commented:
checkstyle pre commit hook으로 코드 스타일에 대해 강제할 수 있도록 하면 좋을 것 같네요~

tlsgkr7416 committed:
Merge remote-tracking branch 'origin/develop' into feature/64/sh
